### PR TITLE
[WIP] Add local blur ("blob") augmentations

### PR DESCRIPTION
**Pull request by mdraw**

### CODE REVIEW

File: elektronn3/data/data_erasing.py [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+                       threshold,
+                       lim_blob_depth,
+                       lim_blob_width,
+                       lim_blob_height,
+                       lim_gauss_diffusion,
+                       verbose=False,
+                       save_path=None,
+                       num_steps_save=None):
+    """ The function takes a batch and applies data erasing.
+    At the beginning the function generates a random number within
```
Comment (mdraw) [inline elektronn3/data/data_erasing.py]: This goes too much into technical detail. It should be more along the lines of "The function randomly generates (random blobs) (...) with the given `probability` value." (not sure if this phrasing is clear/correct, I'm not very confident in my English skills here.)

File: elektronn3/data/data_erasing.py [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+                        intersection.add((k, i, j))
+
+            snippet = batch[batch_indx,
+                            blob.z_min:blob.z_max + 1,
+                            blob.x_min:blob.x_max + 1,
+                            blob.y_min:blob.y_max + 1]
+
+            diffuseness = np.random.randint(low=lim_gauss_diffusion[0],
+                                            high=lim_gauss_diffusion[1],
+                                            dtype=np.int16)
```
Comment (mdraw) [inline elektronn3/data/data_erasing.py]: The `dtype` specification is unnecessary.

File: elektronn3/data/data_erasing.py [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+    batch_volume = batch_depth * batch_width * batch_height
+
+    generator = BlobGenerator(domain=[batch_depth, batch_width, batch_height],
+                              lim_depth=lim_blob_depth,
+                              lim_width=lim_blob_width,
+                              lim_height=lim_blob_height)
+
+    start_time = time.time()
+    threshold.update_value()
+    for batch_indx in range(channels):
```
Comment (mdraw) [inline elektronn3/data/data_erasing.py]: The loop variable name `batch_indx` is misleading. It's a channel index.

File: elektronn3/data/data_erasing.py [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+
+    channels, batch_depth, batch_width, batch_height = batch.shape
+    batch_volume = batch_depth * batch_width * batch_height
+
+    generator = BlobGenerator(domain=[batch_depth, batch_width, batch_height],
+                              lim_depth=lim_blob_depth,
+                              lim_width=lim_blob_width,
+                              lim_height=lim_blob_height)
+
+    start_time = time.time()
```
Comment (mdraw) [inline elektronn3/data/data_erasing.py]: Please delete the timing code and the function call counter, except if you have a good reason for keeping them. Those things can be measured easily with a profiler.

File: elektronn3/data/data_erasing.py [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+    threshold.update_value()
+    for batch_indx in range(channels):
+        erasing_percentage = 0.0
+        intersection = set()
+
+        while erasing_percentage < threshold:
```
Comment (mdraw) [inline elektronn3/data/data_erasing.py]: Some short comments in the following lines would be cool.

File: elektronn3/data/data_erasing.py [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+
+            blob = generator.create_blob()
+
+            for k in range(blob.z_min, blob.z_max + 1):
```
Comment (mdraw) [inline elektronn3/data/data_erasing.py]: (Low priority:) Avoid nested for loops. You can rewrite them with [itertools](https://docs.python.org/3/library/itertools.html#itertools.product).

File: elektronn3/data/data_erasing.py [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+
+    Returns
+    -------
+    None
+    """
+
+    if np.random.rand() > probability:
+        return
+
+    channels, batch_depth, batch_width, batch_height = batch.shape
```
Comment (mdraw) [inline elektronn3/data/data_erasing.py]: As noted above, none of these values are related to a batch, so the names should be updated.

File: elektronn3/data/data_erasing.py [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+    batch_volume = batch_depth * batch_width * batch_height
```
Comment (mdraw) [inline elektronn3/data/data_erasing.py]: (Low priority:) `volume = np.prod(inp.shape[1:])` (given that `batch` is renamed to `inp`.

File: elektronn3/data/data_erasing.py [hunk 9]
```diff
@@ -0,0 +1,372 @@
...
+                for i in range(blob.x_min, blob.x_max + 1):
+                    for j in range(blob.y_min, blob.y_max + 1):
+                        intersection.add((k, i, j))
+
+            snippet = batch[batch_indx,
+                            blob.z_min:blob.z_max + 1,
+                            blob.x_min:blob.x_max + 1,
+                            blob.y_min:blob.y_max + 1]
+
+            diffuseness = np.random.randint(low=lim_gauss_diffusion[0],
```
Comment (mdraw) [inline elektronn3/data/data_erasing.py]: (Low priority:) IMO "diffuseness" is not an optimal name for the sigma parameter of a gaussian blur filter. I'd just call it something like `gaussian_sigma` or `gaussian_std`. But that's just my opinion.

File: elektronn3/data/data_erasing.py [hunk 10]
```diff
@@ -0,0 +1,372 @@
...
+                  blob.y_min:blob.y_max + 1] = snippet
+
+            erased_volume = len(intersection)
+            erasing_percentage = erased_volume / batch_volume
+
+        if verbose:
+            print("erased percentage for channel (%i): %f" %
+                  (batch_indx, erasing_percentage))
+
+        if save_path and num_steps_save:
```
Comment (mdraw) [inline elektronn3/data/data_erasing.py]: This shouldn't be inside the main function. If you still need this, you can put it into an own utility function.

File: elektronn3/data/data_erasing.py [hunk 11]
```diff
@@ -0,0 +1,372 @@
...
+    Returns
+    -------
+    None
+    """
+
+    channels, batch_depth, batch_width, batch_height = batch.shape
+
+    # Check the user's specified blob size.
+    # First entry of each list must be less than the second one
+    if lim_blob_depth[0] >= lim_blob_depth[1]:
```
Comment (mdraw) [inline elektronn3/data/data_erasing.py]: Can you please squash those checks (until line 214) into one or two checks and report all dimensions in case of an error? IMO it would even be fine to replace all these verbose checks with assertions. If any of the assertions fail, you will still know where they occur and can debug them properly.
Another idea would be to use relative size limits in the augmentation function (relative to the input size), instead of specifying fixed sizes. That would have the benefits that all those checks will no longer be needed and you don't have to re-think those sizes when varying input patch sizes in `PatchCreator`.

File: examples/train_unet_neurodata.py [hunk 12]
```diff
@@ -113,6 +133,20 @@
     num_workers=2,
     save_path=save_path,
     schedulers={"lr": lr_sched},
-    cuda_enabled=cuda_enabled
+    cuda_enabled=cuda_enabled,
+    data_erasing_config = erasing_config
 )
-st.train(max_steps)
+
+iterator = st.loader.__iter__()
```
[the remark this answers is not in the document]
Comment (mdraw) [reply examples/train_unet_neurodata.py]: You don't actually need to load an example batch to find out the values that are relevant for your check. You can find the spatial input shape at https://github.com/ELEKTRONN/elektronn3/blob/cacafe99f70d4e452f658b370c2ed6f7f845810c/examples/train_unet_neurodata.py#L81
Please replace the `batch` parameter in `check_data_erasing_config()` with a `patch_shape` parameter (`spatial_shape` could be a better name though) and use the shape from the `data_init_kwargs` dict for it.

File: examples/train_unet_neurodata.py [hunk 13]
```diff
@@ -113,6 +133,20 @@
...
+test_batch, test_target = iterator.__next__()
+
+try:
+    check_data_erasing_config(test_batch.numpy()[0], **erasing_config)
```
Comment (mdraw) [inline examples/train_unet_neurodata.py]: It's less than ideal that this check is performed inside of a user's training script. Once you've addressed the comments above, you should move it to an elektronn3-internal module and perform the check automatically from `PatchCreator` it the new augmentations are used.

File: examples/train_unet_neurodata.py [hunk 14]
```diff
@@ -104,6 +108,22 @@
...
 
+
+threshold = ScheduledVariable(value=0.1,
+                              max_value=0.5,
+                              interval=max_steps,
+                              steps_per_report=None)
+
+erasing_config = {"probability" : 0.75,
+                  "threshold": threshold,
+                  "lim_blob_depth": [5, 15],
```
Comment (mdraw) [inline examples/train_unet_neurodata.py]: The default depth should be halved because of the anisotropy of neuro_data.

File: examples/train_unet_neurodata.py [hunk 15]
```diff
@@ -104,6 +108,22 @@
...
+                  "lim_blob_width": [5, 15],
+                  "lim_blob_height": [5, 15],
+                  "lim_gauss_diffusion": [1, 6],
+                  "verbose": False,
+                  "save_path": "../../trash_bin/",
```
[the remark this answers is not in the document]
Comment (ravil-mobile) [reply examples/train_unet_neurodata.py]: Where do I have to write it though?

[the remark this answers is not in the document]
Comment (mdraw) [reply examples/train_unet_neurodata.py]: Everything related to the current training run should be under `save_path`. But I don't see why writing the "blurred" HDF5 files to disk is even necessary for a training script example.

File: elektronn3/data/blob_generator.py [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+    The user is responsible to pass correct parameters.
+
+    """
+    def __init__(self, domain, lim_depth, lim_width, lim_height):
+        """
+        The construcure initializes all necessary class attributes
+        and prepare an instance to generate blobs
+        Parameters
+        ----------
+        domain - numpy array of integers
```
[the remark this answers is not in the document]
Comment (ravil-mobile) [reply elektronn3/data/blob_generator.py]: Can we postpone that request? I've look at the rest of the code and I haven't seen something like that. Can we at least to decrease the priority of that task?

File: elektronn3/training/trainer.py [hunk 17]
```diff
@@ -151,8 +156,18 @@ def train(self, epochs=1):
...
                 for batch in self.loader:
                     inp, target = batch
+
+                    # copy raw data batch to another location
+                    # to avoid messing up samples within the data loader
+                    # WARNING: possible it slows down the performance
+                    # since the new raw data batch is not pinned
+                    inp = deepcopy(inp)
+
+                    apply_data_erasing(batch=inp.numpy()[0], **self.data_erasing_config)
```
[the remark this answers is not in the document]
Comment (ravil-mobile) [reply elektronn3/training/trainer.py]: I totally agree with that. The problem is: I haven't figured out where exactly I have to insert my augmentation. Whenever I look at the dataloader and stuff which is related to that I go to the "guts" of pytorch which is not a right place to insert my function. Please, propose an idea where exactly it should be

[the remark this answers is not in the document]
Comment (mdraw) [reply elektronn3/training/trainer.py]: As I wrote above, it should be performed directly after the input normalization here: https://github.com/ELEKTRONN/elektronn3/blob/cacafe99f70d4e452f658b370c2ed6f7f845810c/elektronn3/data/cnndata.py#L193

**Comment by mdraw** [comment]
73a554e: Please don't simultaneously rename files and make major changes to them in one commit, because this messes with the git diffs (making all review comments "outdated"). Renaming should be in a separate commit so that git can always understand that files were only moved/renamed, not deleted/created. You can check if renaming was correctly detected by git by checking `git status` between staging and committing.

File: elektronn3/data/data_erasing.py [hunk 18]
```diff
@@ -28,278 +33,296 @@ class FunctionCallsCounter():
...
         Returns
         -------
-        self.value - float
+        None
         """
-        return self.value
+        pass
+
+
+    def check_max_limit(self):
```
Comment (mdraw) [inline elektronn3/data/data_erasing.py]: The name is misleading for a function that both checks and changes a value. Inline this check/modification into the two update functions.

File: elektronn3/data/data_erasing.py [hunk 19]
```diff
@@ -28,278 +33,296 @@ class FunctionCallsCounter():
...
+        """ Checks whether the current parameter value is less
+        than the maximum value specified by the user. If the value exceeds
+        it the parameter variable will be assigned to the maximum one.
+        Returns
+        -------
+        None
+        """
+        if self.value > self.max_value:
```
Comment (mdraw) [inline elektronn3/data/data_erasing.py]: This should be shortened to `self.value = min(self.value, self.max_value)`.

File: elektronn3/data/data_erasing.py [hunk 20]
```diff
@@ -28,278 +33,296 @@ class FunctionCallsCounter():
...
+            self.value = self.max_value
 
-    def __lt__(self, other):
-        return self.value < other
 
-    def __le__(self, other):
-        return self.value <= other
+    def _print_report(self):
+        """ Prints the current value of the parameter on the screen
```
Comment (mdraw) [inline elektronn3/data/data_erasing.py]: *Logs (it's not only printed but also written to a log file)

File: elektronn3/data/data_erasing.py [hunk 21]
```diff
@@ -28,278 +33,296 @@ class FunctionCallsCounter():
...
+        during an iterative process. The function counts number of
```
Comment (mdraw) [inline elektronn3/data/data_erasing.py]: *the number of

File: elektronn3/data/data_erasing.py [hunk 22]
```diff
@@ -28,278 +33,296 @@ class FunctionCallsCounter():
...
+        step() calls and prints information each time when the number
+        of the calls is even with respect to steps_per_report
```
Comment (mdraw) [inline elektronn3/data/data_erasing.py]: *"is divisible by \`\`steps_per_report\`\`."

File: elektronn3/data/data_erasing.py [hunk 23]
```diff
@@ -28,278 +33,296 @@ class FunctionCallsCounter():
...
 
-    def __eq__(self, other):
-        return self.value == other
+        If the used doesn't pass the number of steps_per_report the function
```
Comment (mdraw) [inline elektronn3/data/data_erasing.py]: *user
Avoid negative statements "if not a, then not b". Write it like
"""Logs information every \`\`steps_per_report\`\` steps (if \`\`steps_per_report\`\` is set)."""

File: elektronn3/data/data_erasing.py [hunk 24]
```diff
@@ -28,278 +33,296 @@ class FunctionCallsCounter():
...
+            the parameter value at the beginning of an scheduled process
         max_value - float
+            the parameter value at the end of an scheduled process
+        growth_type - str
+            type of growth: "lin" - linear; "exp" - exponential
         interval - int
+            number of steps along which the parameter value has to be
+            increased from the initial value to the maximal one
         steps_per_report - int
+            number of step between information update on the screen
```
Comment (mdraw) [inline elektronn3/data/data_erasing.py]: The special case `steps_per_report = None` should be mentioned here, especially since it's the default.

File: elektronn3/data/data_erasing.py [hunk 25]
```diff
@@ -28,278 +33,296 @@ class FunctionCallsCounter():
...
-
-    If the user doesn't specify the target value or the interval the variable
-    works as a constant
+    If the user doesn't specify the target value or the interval,
+    the parameter works as a constant
 
     """
-    def __init__(self, value, max_value=None, interval=None, steps_per_report=None):
+
+    logger = logging.getLogger('elektronn3log')
```
Comment (mdraw) [inline elektronn3/data/data_erasing.py]: Don't tie the logger to a specific class. Use a module-level logger instead, like in all other modules of elektronn3 that perform logging.

File: elektronn3/data/data_erasing.py [hunk 26]
```diff
@@ -28,278 +33,296 @@ class FunctionCallsCounter():
...
 
-    def __eq__(self, other):
-        return self.value == other
+        If the used doesn't pass the number of steps_per_report the function
+        doesn't print the information
+        Returns
+        -------
+        None
+        """
+        if self.steps_per_report:
```
Comment (mdraw) [inline elektronn3/data/data_erasing.py]: `if self.steps_per_report` suggests that it's a bool.
This should be
`if self.steps_per_report is not None`.
`None` checks should always be explicit.

**Comment by mdraw** [comment]
We can re-do the renaming of data_erasing.py later, just before merging the PR (I am a little scared that renaming it now will mess up the review again).
